interface/usergroup/mfa_totp.php

```diff
@@ -0,0 +1,218 @@
+
+        function doregister(step, error) {
+            var f = document.forms[0];
+            f.action.value = step;
+            if (error) {
+                f.error.value = error;
+            }
+            f.action.value = step;
+            top.restoreSession();
+            f.submit();
+        }
+
+        function docancel() {
+            var redirectUrl = 'mfa_registrations.php';
+            window.location.href = 'mfa_registrations.php';
+        }
+
+        function dodelete() {
+            var f = document.forms[0];
+            f.action = 'mfa_registrations.php';
+            doregister('delete');
+        }
+    </script>
+</head>
+<body class="body_top">
+<form method='post' class="form-horizontal" action='mfa_totp.php' onsubmit='return top.restoreSession()'>
+    <input type="hidden" name="csrf_token_form" value="<?php echo attr(collectCsrfToken()); ?>" />
+
+
+    <?php
+    // If current step is reg1 or reg2, display the header
+    if ($action == 'reg1' || $action == 'reg2') { ?>
+        <div class="container">
+            <div class="row">
+                <div class="col-xs-12">
+                    <div class="page-header">
+                        <h3><?php echo xlt('Register App Based TOTP'); ?></h3>
+                    </div>
+                </div>
+            </div>
+    <?php } ?>
+
+    <?php
+    // step 1 is to verify the password
+    if ($action == 'reg1') {
+        $error = (isset($_GET["error"])) ? $_GET["error"] : false;
+        ?>
+            <div class="row">
+                <div class="col-xs-12">
+                    <?php if ($error == "auth") { ?>
+                        <div class="alert alert-danger login-failure m-1">
+                            Invalid password
+                        </div>
+                    <?php } ?>
+                    <p>In order to register your device, please provide your password</p>
+                    <table cellspacing="5">
+                        <tr>
+                            <td>
+                                <label for="clearPass"><?php echo xlt('Password:'); ?>
+                            </td>
+                            <td>
+                                <input type="password" class="form-control" id="clearPass" name="clearPass" placeholder="<?php echo xla('Password:'); ?>" >
+                            </td>
+                        </tr>
+                        <tr>
+                            <td></td>
+                            <td>
+                                <input type="button" value="Submit" onclick="doregister('reg2')" />
+                                <input type="button" value="Cancel" onclick="docancel()" />
+                            </td>
+                        </tr>
+                    </table>
+                </div>
+            </div>
+        </div>
+<?php
+    // step 2 is to validate password and display qr code
+    } elseif ($action == 'reg2') {
+
+        // Validate user password
```

Comment:
Can't we use this function:
https://github.com/openemr/openemr/blob/master/library/authentication/login_operations.php#L27

Reply:
I'll try to use this function again. In my development, I think I was having issues with it because in the function call collectIpAddresses, $_SERVER['HTTP_X_FORWARDED_FOR'] is not being found. I'll try to integrate this method again and reach out if I have issues.